### PR TITLE
Fixed #23745 -- Improved migrate speed

### PR DESCRIPTION
**Pull request by claudep**
Of course, commit c7ea633 will not be eventually included, it only serves to measure performance.

### CODE REVIEW

File: django/core/management/commands/migrate.py [hunk 1]
```diff
@@ -201,14 +201,17 @@ def handle(self, *args, **options):
...
+                elapsed = time.time() - self.start
                 if fake:
-                    self.stdout.write(self.style.MIGRATE_SUCCESS(" FAKED"))
+                    self.stdout.write(self.style.MIGRATE_SUCCESS(" FAKED (%s)" % elapsed))
                 else:
-                    self.stdout.write(self.style.MIGRATE_SUCCESS(" OK"))
+                    self.stdout.write(self.style.MIGRATE_SUCCESS(" OK (%s)" % elapsed))
             elif action == "unapply_start":
                 self.stdout.write("  Unapplying %s..." % migration, ending="")
                 self.stdout.flush()
```
Comment (claudep) [inline django/core/management/commands/migrate.py]: Would it be of any use to keep the elapsed time display when verbosity > 1?

Comment (timgraham) [inline django/core/management/commands/migrate.py]: could be interesting I suppose

Comment (MarkusH) [inline django/core/management/commands/migrate.py]: I thought about that as well. The only question I have is, do we already show it for a verbosity of 2 or only 3? I normally to run everything with `-v 2`, but the elapsed time is nothing that'd be useful unless I'm debugging something, which I would run with `-v 3` anyway.

Comment (claudep) [inline django/core/management/commands/migrate.py]: The problem with requiring `-v 3` is that `create_test_db` is substracting 1 to verbosity when calling `migrate` and `runtests.py` doesn't accept verbosity greater than 3.

Comment (MarkusH) [inline django/core/management/commands/migrate.py]: Let's go with `-v 2` then.

File: django/db/migrations/operations/fields.py [hunk 2]
```diff
@@ -38,11 +38,12 @@ def state_forwards(self, app_label, state):
         else:
             field = self.field
         state.models[app_label, self.model_name.lower()].fields.append((self.name, field))
+        state.reload_model(app_label, self.model_name.lower())
```
Comment (MarkusH) [inline django/db/migrations/operations/fields.py]: You don't need the `lower()` here.

File: django/db/migrations/operations/fields.py [hunk 3]
```diff
@@ -94,16 +95,17 @@ def state_forwards(self, app_label, state):
             if name != self.name:
                 new_fields.append((name, instance))
         state.models[app_label, self.model_name.lower()].fields = new_fields
+        state.reload_model(app_label, self.model_name.lower())
```
Comment (MarkusH) [inline django/db/migrations/operations/fields.py]: Neither do you need `lower()` here.

File: django/db/migrations/state.py [hunk 4]
```diff
@@ -26,80 +29,63 @@ class ProjectState(object):
...
+    def add_model(self, model_state):
+        app_label, model_name = model_state.app_label, model_state.name.lower()
+        self.models[(app_label, model_name)] = model_state
+        if 'apps' in self.__dict__:  # hasattr would cache the property
+            self.reload_model(app_label, model_name)
+
+    def remove_model(self, app_label, model_name):
+        model_name = model_name.lower()
+        del self.models[app_label, model_name]
+        if 'apps' in self.__dict__:
```
Comment (MarkusH) [inline django/db/migrations/state.py]: Can you copy the comment from `add_model` here.

File: django/db/migrations/state.py [hunk 5]
```diff
@@ -26,80 +29,63 @@ class ProjectState(object):
...
+            self.apps.unregister_model(app_label, model_name)
+
+    def reload_model(self, app_label, model_name):
+        if 'apps' in self.__dict__:
```
Comment (MarkusH) [inline django/db/migrations/state.py]: Can you copy the comment from `add_model` here.

File: django/db/migrations/state.py [hunk 6]
```diff
@@ -26,80 +29,63 @@ class ProjectState(object):
...
+            # Get relations before reloading the models, as _meta.apps may change
+            model_name = model_name.lower()
+            try:
+                related_old = {
+                    f.model for f in
+                    self.apps.get_model(app_label, model_name)._meta.get_all_related_objects()}
```
Comment (MarkusH) [inline django/db/migrations/state.py]: I'd put the trailing `}` in a new line.

**Comment by MarkusH** [comment]
Two more comments:

- Could you stick to the new naming "old_state" and "project_state" instead of "project_state" and "new_state" in `django.db.migrations.migration.Migration.unapply()`
- I dropped a few more `.lower()` calls:
```patch
diff --git a/django/db/migrations/operations/fields.py b/django/db/migrations/operations/fields.py
index a10e0d1..b6a66f0 100644
--- a/django/db/migrations/operations/fields.py
+++ b/django/db/migrations/operations/fields.py
@@ -34,7 +34,7 @@ class AddField(Operation):
         else:
             field = self.field
         state.models[app_label, self.model_name.lower()].fields.append((self.name, field))
-        state.reload_model(app_label, self.model_name.lower())
+        state.reload_model(app_label, self.model_name)
 
     def database_forwards(self, app_label, schema_editor, from_state, to_state):
         to_model = to_state.apps.get_model(app_label, self.model_name)
@@ -87,7 +87,7 @@ class RemoveField(Operation):
             if name != self.name:
                 new_fields.append((name, instance))
         state.models[app_label, self.model_name.lower()].fields = new_fields
-        state.reload_model(app_label, self.model_name.lower())
+        state.reload_model(app_label, self.model_name)
 
     def database_forwards(self, app_label, schema_editor, from_state, to_state):
         from_model = from_state.apps.get_model(app_label, self.model_name)
@@ -140,7 +140,7 @@ class AlterField(Operation):
         state.models[app_label, self.model_name.lower()].fields = [
             (n, field if n == self.name else f) for n, f in state.models[app_label, self.model_name.lower()].fields
         ]
-        state.reload_model(app_label, self.model_name.lower())
+        state.reload_model(app_label, self.model_name)
 
     def database_forwards(self, app_label, schema_editor, from_state, to_state):
         to_model = to_state.apps.get_model(app_label, self.model_name)
@@ -206,7 +206,7 @@ class RenameField(Operation):
                     [self.new_name if n == self.old_name else n for n in together]
                     for together in options[option]
                 ]
-        state.reload_model(app_label, self.model_name.lower())
+        state.reload_model(app_label, self.model_name)
 
     def database_forwards(self, app_label, schema_editor, from_state, to_state):
         to_model = to_state.apps.get_model(app_label, self.model_name)
diff --git a/django/db/migrations/operations/models.py b/django/db/migrations/operations/models.py
index 4724094..3642b3d 100644
--- a/django/db/migrations/operations/models.py
+++ b/django/db/migrations/operations/models.py
@@ -88,7 +88,7 @@ class DeleteModel(Operation):
         )
 
     def state_forwards(self, app_label, state):
-        state.remove_model(app_label, self.name.lower())
+        state.remove_model(app_label, self.name)
 
     def database_forwards(self, app_label, schema_editor, from_state, to_state):
         model = from_state.apps.get_model(app_label, self.name)
@@ -133,7 +133,7 @@ class RenameModel(Operation):
         # Rename the model
         state.models[app_label, self.new_name.lower()] = state.models[app_label, self.old_name.lower()]
         state.models[app_label, self.new_name.lower()].name = self.new_name
-        state.remove_model(app_label, self.old_name.lower())
+        state.remove_model(app_label, self.old_name)
         # Repoint the FKs and M2Ms pointing to us
         for related_object in (related_objects + related_m2m_objects):
             # Use the new related key for self referential related objects.
@@ -152,7 +152,7 @@ class RenameModel(Operation):
                 new_fields.append((name, field))
             state.models[related_key].fields = new_fields
             state.reload_model(*related_key)
-        state.reload_model(app_label, self.new_name.lower())
+        state.reload_model(app_label, self.new_name)
 
     def database_forwards(self, app_label, schema_editor, from_state, to_state):
         new_model = to_state.apps.get_model(app_label, self.new_name)
@@ -219,7 +219,7 @@ class AlterModelTable(Operation):
 
     def state_forwards(self, app_label, state):
         state.models[app_label, self.name.lower()].options["db_table"] = self.table
-        state.reload_model(app_label, self.name.lower())
+        state.reload_model(app_label, self.name)
 
     def database_forwards(self, app_label, schema_editor, from_state, to_state):
         new_model = to_state.apps.get_model(app_label, self.name)
@@ -271,7 +271,7 @@ class AlterUniqueTogether(Operation):
     def state_forwards(self, app_label, state):
         model_state = state.models[app_label, self.name.lower()]
         model_state.options[self.option_name] = self.unique_together
-        state.reload_model(app_label, self.name.lower())
+        state.reload_model(app_label, self.name)

     def database_forwards(self, app_label, schema_editor, from_state, to_state):
         new_model = to_state.apps.get_model(app_label, self.name)
@@ -315,7 +315,7 @@ class AlterIndexTogether(Operation):
     def state_forwards(self, app_label, state):
         model_state = state.models[app_label, self.name.lower()]
         model_state.options[self.option_name] = self.index_together
-        state.reload_model(app_label, self.name.lower())
+        state.reload_model(app_label, self.name)

     def database_forwards(self, app_label, schema_editor, from_state, to_state):
         new_model = to_state.apps.get_model(app_label, self.name)
@@ -356,7 +356,7 @@ class AlterOrderWithRespectTo(Operation):
     def state_forwards(self, app_label, state):
         model_state = state.models[app_label, self.name.lower()]
         model_state.options['order_with_respect_to'] = self.order_with_respect_to
-        state.reload_model(app_label, self.name.lower())
+        state.reload_model(app_label, self.name)

     def database_forwards(self, app_label, schema_editor, from_state, to_state):
         to_model = to_state.apps.get_model(app_label, self.name)
@@ -420,7 +420,7 @@ class AlterModelOptions(Operation):
         for key in self.ALTER_OPTION_KEYS:
             if key not in self.options and key in model_state.options:
                 del model_state.options[key]
-        state.reload_model(app_label, self.name.lower())
+        state.reload_model(app_label, self.name)

     def database_forwards(self, app_label, schema_editor, from_state, to_state):
         pass
```

**Comment by MarkusH** [comment]
If the tests still pass, I think this PR is ready to go. Thank you @claudep !

**Pull request merged**